### PR TITLE
39 remove tester and update test transformer to target as pect

### PR DESCRIPTION
**Pull request by peterjah**
In this PR:
- upgrade assemblyscript to 0.24.1
- upgrade as-pect to 8.0.0
- refactor of file2byteArray transformer => use Typescript and visitor-as, proper packaging for @massalabs/as-transformer
- change testTable transformer (checksForEachLineThatThe) to work with as-pect and included in new transformer package
- remove @massalabs/tester, i will mark it as deprecated in npm
- remove unused dependencies (as-big, as-console and as-wasi)

### CODE REVIEW

File: tester/assembly/tsconfig.json [hunk 1]
```diff
@@ -2,5 +2,5 @@
   "extends": "assemblyscript/std/assembly.json",
   "include": [
     "./**/*.ts"
-  ]
-}
\ No newline at end of file
+]
```
Comment (Thykof) [inline tester/assembly/tsconfig.json:5]: I think we should keep those 2 spaces

File: transformer/assembly/__tests__/file2ByteArray.spec.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+export function strToStaticArray(str: string): StaticArray<u8> {
```
Comment (Thykof) [inline transformer/assembly/__tests__/file2ByteArray.spec.ts:1]: Is there a need to export this function?

```suggestion
function strToStaticArray(str: string): StaticArray<u8> {
```

File: transformer/src/file2ByteArray.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import {
+  CallExpression,
+  Expression,
+  StringLiteralExpression,
+} from 'assemblyscript/dist/assemblyscript.js';
+import {SimpleParser} from 'visitor-as';
+
+import * as fs from 'fs';
+
+export class File2ByteArray {
```
Comment (gregLibert) [inline transformer/src/file2ByteArray.ts:10]: You should document it for the next dev ;-)

File: transformer/types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
+/// <reference types="assemblyscript/std/assembly/index" />
+
+declare function fileToByteArray(filePath: string): StaticArray<u8>;
```
Comment (gregLibert) [inline transformer/types/index.d.ts:3]: Don't you think that we document all that ?

File: transformer/types/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
+/// <reference types="assemblyscript/std/assembly/index" />
+
```
Comment (Thykof) [inline transformer/types/index.d.ts:2]: is it possible to add something explaining how to use this file?

File: transformer/src/testTable.ts [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+ * - is, isNot,
+ * - isTrue, isFalse,
+ *
+ * If the criterion is not in the list,
+ * the raw string value is taken.
+ *
+ * @param {string} criterion
+ * @return {string}
+ */
+export function generateIfExpression(criterion: string) {
```
Comment (Thykof) [inline transformer/src/testTable.ts:22]: why export this function? and we may want to put it in the class below

File: transformer/src/testTable.ts [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+ *
+ * @param {string} name - test friendly name
+ * @param {string} got - got expression
+ * @param {string} expect - expect expression
+ * @param {string} ifExpression
+ * @param {bool} needWant - include want in test ?
+ * @param {bool} continueOnFailure - should the test failure stop the test set
+ * @return {Object}
+ */
+export function generateTest(
```
Comment (Thykof) [inline transformer/src/testTable.ts:55]: why export this function? and we may want to put it in the class below

File: transformer/src/testTable.ts [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+/**
+ * Hydrates the template by replacing the tokens (arg[0-9]+) with actual values.
+ *
+ * @param {string} template - template
+ * @param {Array} value - array of values to populate template with
+ * @param {number} iValue - current index of values
+ * @return {Object}
+ */
+function hydrateTemplate(template: string, value: any, iValue: number) {
```
Comment (Thykof) [inline transformer/src/testTable.ts:91]: why don't put this function in the class bellow?

**Review by Thykof — APPROVED** [review]

File: transformer/src/testTable.ts (hunk 6 above)
Comment (peterjah) [reply transformer/src/testTable.ts:22]: true!

File: transformer/types/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+): void;
+
+declare enum onFailure {
+  Continue,
+  ExitSet,
+}
+
+declare const is = '';
+declare const isNot = '';
+declare const isTrue = '';
```
[the remark this answers is not in the document]
Comment (peterjah) [reply transformer/types/index.d.ts:27]: its just a typings file. only used for your editor don't complain about it beeing not declared

File: transformer/src/testTable.ts [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+  let instanciation = template;
```
Comment (SebastienValla) [inline transformer/src/testTable.ts:92]:
```suggestion
  let instantiation = template;
```

File: transformer/src/testTable.ts [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+  while (instanciation.search(/arg[0-9]/) > -1) {
+    instanciation = instanciation.replace(/arg[0-9]+/, value[iValue]);
+    iValue++;
+  }
+
+  return {
+    instanciation,
+    iValue,
+  };
+}
```
Comment (SebastienValla) [inline transformer/src/testTable.ts:103]:
```suggestion
  while (instantiation.search(/arg[0-9]/) > -1) {
    instantiation = instantiation.replace(/arg[0-9]+/, value[iValue]);
    iValue++;
  }

  return {
    instantiation,
    iValue,
  };
}
```

**Review by SebastienValla — CHANGES_REQUESTED** [review]

File: transformer/src/testTable.ts [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+        expectExpr.replace(/['`]/g, ''),
+        ifExpr,
+        needWant,
+      );
+
+      expr += test.replace(/\n/g, '\n  '); // adds static indentation
+    }
+    expr += `\n});`;
+
+    const processedNode = SimpleParser.parseExpression(expr);
```
Comment (gregLibert) [inline transformer/src/testTable.ts:96]: Your terminology is misleading: your are creating a new node and the processedNode is the current node variable.

File: transformer/src/testTable.ts [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+    return RangeTransform.visit(processedNode, node);
```
Comment (gregLibert) [inline transformer/src/testTable.ts:97]: Why are you calling RangeTransform ?

**Review by gregLibert — CHANGES_REQUESTED** [review]

File: package.json [hunk 14]
```diff
@@ -33,15 +33,15 @@
...
+    "@as-pect/cli": "^8.0.0",
+    "@assemblyscript/loader": "^0.24.1",
     "@envy-as/cli": "https://github.com/jtenner/envy",
     "@typescript-eslint/eslint-plugin": "^5.18.0",
     "@typescript-eslint/parser": "^5.18.0",
     "as-big": "^0.2.0",
     "as-console": "^6.0.2",
     "as-wasi": "^0.5.1",
-    "assemblyscript": "^0.21.2",
+    "assemblyscript": "^0.24.1",
```
[the remark this answers is not in the document]
Comment (peterjah) [reply package.json:46]: it has been released yesterday :smile:
Should i take it ? its very fresh..

File: as-pect.asconfig.json [hunk 15]
```diff
@@ -2,13 +2,13 @@
...
       "lib": ["./node_modules/@as-covers/assembly/index.ts"],
-      "transform": ["@as-covers/transform", "@as-pect/transform"]
+      "transform": ["@as-covers/transform"]
     },
     "noCoverage": {
-      "transform": ["@as-pect/transform"]
     }
   },
   "options": {
+    "transform": ["@as-pect/transform", "@massalabs/as-transformer"],
```
[the remark this answers is not in the document]
Comment (peterjah) [reply as-pect.asconfig.json:11]: we are lucky, it works in any order !

File: package.json [hunk 16]
```diff
@@ -22,20 +22,27 @@
     "registry": "https://registry.npmjs.org/"
   },
   "scripts": {
-    "test": "asp --verbose",
+    "clean": "rimraf \"./transformer/dist\"",
+    "install:transformer": "npm install --prefix ./transformer",
+    "build:transformer": "tsc -p transformer",
```
[the remark this answers is not in the document]
Comment (peterjah) [reply package.json:27]: i don't get your point! its actually compiling the transformer directory which is typescript

File: package.json [hunk 17]
```diff
@@ -22,20 +22,27 @@
...
+    "build": "npm run clean && npm run build:transformer",
+    "test": "asp",
     "test:ci": "asp --summary",
-    "prettier": "as-prettier --check assembly \"./**/assembly\"",
-    "prettier:fix": "as-prettier --write assembly \"./**/assembly\"",
+    "asPrettier": "as-prettier \"**/assembly/**/*.ts\"",
```
[the remark this answers is not in the document]
Comment (peterjah) [reply package.json]: you are reviewing an outdated code ?

File: package.json [hunk 18]
```diff
@@ -22,20 +22,27 @@
...
+    "prettier": "npm run asPrettier -- --check",
+    "prettier:fix": "npm run asPrettier -- --write",
     "lint": "eslint .",
     "lint:fix": "eslint --fix .",
     "fmt:check": "npm run prettier && npm run lint",
     "fmt": "npm run prettier:fix && npm run lint:fix",
+    "prepublish": "npm run install:transformer",
```
[the remark this answers is not in the document]
Comment (peterjah) [reply package.json:37]: like this name NOT indicates :smile: its a hook wich is executed after npm install. so npm install will also trigger the install of the transformer package

File: transformer/types/index.d.ts (hunk 4 above)
Comment (peterjah) [reply transformer/types/index.d.ts:3]: I added some stuffs about that in the readme

File: transformer/src/testTable.ts (hunk 13 above)
Comment (peterjah) [reply transformer/src/testTable.ts:97]: it is used to preserve some attributes from the original node. returning directly the created node wont work

Comment (gregLibert) [reply transformer/src/testTable.ts]: Don't you think we should comment it ?

File: package.json (hunk 14 above)
[the remark this answers is not in the document]
Comment (gregLibert) [reply package.json:46]: ok understood.

File: package.json [hunk 19]
```diff
@@ -22,20 +22,27 @@
     "registry": "https://registry.npmjs.org/"
   },
   "scripts": {
-    "test": "asp --verbose",
+    "clean": "rimraf \"./transformer/dist\"",
+    "install:transformer": "npm install --prefix ./transformer",
```
[the remark this answers is not in the document]
Comment (gregLibert) [reply package.json:26]: okidoki

File: package.json (hunk 18 above)
[the remark this answers is not in the document]
Comment (gregLibert) [reply package.json:37]: Cool. Maybe we should document that to remove the "magic"!

**Review by Thykof — APPROVED** [review]

**Review by SebastienValla — APPROVED** [review]

**Review by gregLibert — APPROVED** [review]
🙈

**Pull request merged**